File: js/25343.js
const fs = require('fs');

const [[N], ...arr] = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

const dp = Array(N)
  .fill()
  .map(() => Array(N).fill(0));

dp[0][0] = 1;

for (let i = 0; i < N; i++) {
  dp[0][i] = 1;
  dp[i][0] = 1;
  for (let j = 0; j < i; j++) {
    if (arr[0][j] < arr[0][i]) dp[0][i] = Math.max(dp[0][i], dp[0][j] + 1);
  }
  for (let j = 0; j < i; j++) {
    if (arr[j][0] < arr[j][0]) dp[i][0] = Math.max(dp[i][0], dp[j][0] + 1);
  }
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    dp[i][j] = 1;
    for (let a = 0; a <= i; a++) {
      for (let b = 0; b <= j; b++) {
        if (arr[a][b] < arr[i][j]) dp[i][j] = Math.max(dp[i][j], dp[a][b] + 1);
      }
    }
  }
}

let ans = 0;

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    ans = Math.max(ans, dp[i][j]);
  }
}

console.log(ans);

File: c++/25343.cpp
#include <iostream>

using namespace std;

int dp[101][101];  // dp[i][j] : (i ,j) 위치에 도달했을 때 지나온 최대 칸 수
int arr[101][101];

int main() {
  freopen("example.txt", "r", stdin);

  int n;
  cin >> n;

  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= n; j++) {
      cin >> arr[i][j];
    }
  }

  dp[1][1] = 1;

  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= n; j++) {
      for (int a = 0; a <= i; a++) {
        for (int b = 0; b <= j; b++) {
          if (arr[a][b] < arr[i][j])
            dp[i][j] = max(dp[i][j], dp[a][b] + 1);
          else if (arr[a][b] == arr[i][j])
            dp[i][j] = max(dp[i][j], dp[a][b]);
        }
      }
    }
  }

  int ans = 0;
  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= n; j++) {
      cout << dp[i][j] << ' ';
      if (dp[i][j] > ans) ans = dp[i][j];
    }
    cout << '\n';
  }

  cout << ans;

  return 0;
}